MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c

```diff
@@ -13,30 +13,119 @@ MU_CHANGE: new file
+  DEVICE_STATE_MANUFACTURING_MODE_STRING,
+  DEVICE_STATE_DEVELOPMENT_BUILD_ENABLED_STRING,
+  DEVICE_STATE_SOURCE_DEBUG_ENABLED_STRING,
+  DEVICE_STATE_UNDEFINED_STRING,
+  DEVICE_STATE_UNIT_TEST_MODE_STRING,
+};
+
 /**
-Function to Get current device state
-@retval the current DEVICE_STATE
-**/
-DEVICE_STATE
+ * Get String that represents the device's current insecure states, using the platform defined set of
+ * insecure device states. This can be used to measure insecure device states into the TPM or perform
+ * other required platform actions when the device enters an insecure state.
+ *
+ * @param[out] Buffer - Buffer to store the string
+ * @param[in] MaxSize - Maximum size of the buffer
+ * @retval The The number of characters in the string, including the null character.
+ */
+UINTN
 EFIAPI
-GetDeviceState (
+GetInsecureDeviceStateString (
+  OUT CHAR8  *Buffer,
+  IN UINTN   MaxSize
   )
 {
-  DEVICE_STATE  DevState = PcdGet32 (PcdDeviceStateBitmask);
+  CHAR8         *String;
+  DEVICE_STATE  BitmaskFromIndex;
+  DEVICE_STATE  CurrentDeviceState         = GetDeviceState ();
+  DEVICE_STATE  InsecureDeviceStateSetting = GetInsecureDeviceStateSetting ();
+
+  UINT32  NumberOfDeviceStates = sizeof (DEVICE_STATE_STRINGS) / sizeof (DEVICE_STATE_STRINGS[0]);
+
+  //
+  // Check standard device states
+  //
+  for (UINT32 i = 0; i < NumberOfDeviceStates; i++) {
+    BitmaskFromIndex = 1 << i;
+    if ((InsecureDeviceStateSetting & BitmaskFromIndex != 0) &&
+        (CurrentDeviceState & BitmaskFromIndex != 0))
+    {
+      //
+      // An insecure device state is detected. Append to the insecure device state string.
+      //
+      String = DEVICE_STATE_STRINGS[i];
+      AsciiStrnCatS (Buffer, MaxSize, String, AsciiStrLen (String));
+    }
+  }
 
-  return DevState;
+  //
+  // Check OEM-defined device states
+  //
+  for (UINT32 i = DEVICE_STATE_PLATFORM_MODE_INDEX_START; i <= DEVICE_STATE_PLATFORM_MODE_INDEX_END; i++) {
+    BitmaskFromIndex = 1 << i;
+    if ((InsecureDeviceStateSetting & BitmaskFromIndex != 0) &&
+        (CurrentDeviceState & BitmaskFromIndex != 0))
+    {
+      //
+      // An OEM insecure device state is detected. Append to the insecure device state string.
+      //
+      String = DEVICE_STATE_PLATFORM_MODE_STRING;
+      AsciiStrnCatS (Buffer, MaxSize, String, AsciiStrLen (String));
+      break;
+    }
+  }
+
+  return AsciiStrnLenS (Buffer, MaxSize) + 1;
+}
+
+/**
+ * Get the platform defined set of insecure device states. This can be used to measure insecure device
+ * states into the TPM or perform other required platform actions when the device enters an insecure state.
+ *
+ * @retval The bitmask of insecure device states as defined by the platform.
+ */
+DEVICE_STATE
+EFIAPI
+GetInsecureDeviceStateSetting (
+  )
+{
+  return FixedPcdGet32 (PcdInsecureDeviceState);
 }
 
 /**
-Function to Add additional bits to the device state
+ * Check if a specific bitmask is set in the device state

```

Comment:
Is a function definition missing between these comments?

```c
/**
 * Check if a specific bitmask is set in the device state
 *
 * @param[in] CurrentDeviceState - the current device state
 * @param[in] DeviceStateBitmask - the bitmask to check
 * @retval TRUE if all the bits in the bitmask are set, FALSE otherwise
 */
```

Reply:
Removed the function earlier and this was missed. Cleaned up now.